src/main/java/org/ballerinax/datamapper/DataMapperNodeVisitor.java

```diff
@@ -0,0 +1,208 @@
+
+    public DataMapperNodeVisitor() {
+        this.recordTypes = new HashMap<String, String>();
+        this.clientMap = new HashMap<>();
+    }
+
+    public HashMap<String, String> getRecordTypes() {
+        return recordTypes;
+    }
+
+    public HashMap<String, Map<String, FunctionRecord>> getClientMap() {
+        return clientMap;
+    }
+
+    public void setModule(Module module) {
+        ModuleCompilation compilation = module.getCompilation();
+        this.model = compilation.getSemanticModel();
+    }
+
+    private String getFieldTypes(Map<String, RecordFieldSymbol> fieldSymbolMap) {
+        Iterator<String> iterator = fieldSymbolMap.keySet().iterator();
+        Map<String, String> fieldSymbols = new HashMap<>();
+        String serialized = null;
+        while (iterator.hasNext()) {
+            String fieldName = iterator.next();
+            String fieldType = fieldSymbolMap.get(fieldName).typeDescriptor().signature();
+            fieldSymbols.put(fieldName, fieldType);
+        }
+        try {
+            serialized = new ObjectMapper().writeValueAsString(fieldSymbols);
+        } catch (JsonProcessingException e) {
+            throw new DataMapperException(e);
+        }
+        return serialized;
+    }
+
+    @Override
+    public void visit(TypeDefinitionNode typeDefinitionNode) {
+        if (typeDefinitionNode.typeDescriptor().kind() == SyntaxKind.RECORD_TYPE_DESC) {
+            Optional<Symbol> recordSymbolOpt = this.model.symbol(typeDefinitionNode);
+            if (recordSymbolOpt.isPresent()) {
+                Symbol recordSymbol = recordSymbolOpt.get();
+                Optional<String> recordNameOpt = recordSymbol.getName();
+                if (recordNameOpt.isPresent()) {
+                    String recordName = recordNameOpt.get();
+                    String recordSignature = recordSymbol.getModule().get().id().toString();
+                    recordName = recordSignature + ":" + recordName;
+                    Map<String, RecordFieldSymbol> fieldSymbolMap = ((RecordTypeSymbol) ((TypeDefinitionSymbol)
+                            recordSymbol).typeDescriptor()).fieldDescriptors();
+                    String serialized = getFieldTypes(fieldSymbolMap);
+                    serialized = "{\"" + recordName + "\":" + serialized + "}";
+                    this.recordTypes.put(recordName, serialized);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void visit(ClassDefinitionNode classDefinitionNode) {
+        NodeList<Token> classTypeQualifiers = classDefinitionNode.classTypeQualifiers();
+        for (Token classTypeQualifier : classTypeQualifiers) {
+            if (classTypeQualifier.text().equals("client")) {
+                if (model.symbol(classDefinitionNode).isPresent()) {
+                    Symbol symbol = model.symbol(classDefinitionNode).get();
+                    if (symbol.getName().isEmpty()) {
+                        continue;
+                    }
```

Comment:
Can first lookup the class symbol first and check if it contains the `CLIENT` qualifier. Also, better not to lookup the symbol multiple times as it will result in a tree traversal each time.

Reply:
Done